### PR TITLE
Use XDG dirs by default on Linux

### LINKED ISSUE #453: Data folder placement: use XDG dirs on Linux
On Linux the Audacity data folder is on:
~/.audacity-data

Most likely it should be:
~/.config/audacity

### PR DESCRIPTION
**Pull request by RuRo**
This PR is an updated version of #831 which itself is an updated version of #632

Fixes #453 and resolves: https://bugzilla.audacityteam.org/show_bug.cgi?id=2201

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by RuRo** [comment]
@LWinterberg @crsib is there anything left on my side, that is blocking this PR or are we just waiting for someone to review this?

**Comment by LWinterberg** [comment]
I think the main blocker for this was the weekend :) 

The review is already inside our kanban board, so crsib will look at it soon

File: libraries/lib-files/PathList.cpp
Comment (crsib) [inline libraries/lib-files/PathList.cpp:24]: Probably it should go to `#if defined(__WXGTK__)` as well

File: src/AboutDialog.cpp
Comment (crsib) [inline src/AboutDialog.cpp:615]: I think we could leave "Settings folder" here. What do you think, @Paul-Licameli?

File: libraries/lib-files/PathList.cpp
Comment (RuRo) [reply libraries/lib-files/PathList.cpp:24]: Will do.

File: src/AboutDialog.cpp
Comment (RuRo) [reply src/AboutDialog.cpp:615]: Okay.

File: libraries/lib-files/FileNames.cpp
Comment (RuRo) [reply libraries/lib-files/FileNames.cpp:261]: I am not sure, if I follow? `GetUserDataDir()` will return `~/.local/share`, not `~/.config`. +1 on using `AUDACITY_NAME` instead of a hardcoded string.

Comment (crsib) [reply libraries/lib-files/FileNames.cpp:261]: Docs are confusing: `Only use this method if you have a single configuration file to put in this directory, otherwise [GetUserDataDir()](https://docs.wxwidgets.org/3.1/classwx_standard_paths.html#a5b9d1b1addc3e4ce30ccb3817cbbe19b) is more appropriate as the latter adds appinfo to the path, unlike this function.` https://docs.wxwidgets.org/3.1/classwx_standard_paths.html#a0561d70f7da648e10b25bf9679309da3

Comment (RuRo) [reply libraries/lib-files/FileNames.cpp:261]: I agree, that the docs are confusing. I am currently rewriting the implementation based on your `enum DirTarget` idea.

Actually the `wxwidgets` `GetUserWhateverDir` methods don't map to corresponding `XDG_WHATEVER_HOME` directories, so I decided to just implement the correct XDG Base Directory logic for unix and keep the windows case exactly as it currently is.

More info later. Btw, are we sure that `__WXGTK__` is the correct preprocessor macro to distinguish unix? I just kept it from the original PR, but I am not 100% sure.

Comment (crsib) [reply libraries/lib-files/FileNames.cpp:261]: > are we sure that __WXGTK__ 

We use it for now, at least, so it's good to go!

**Comment by RuRo** [comment]
@crsib PTAL. I've implemented the 4 XDG Base Directories. Currently, all 4 directories default back to `~/.audacity-data` if it already exists and to `GetUserDataDir` on Windows.

Also, I completely removed the `FileLayout_XDG` logic since it would only affect `Config` and we would still need to implement the XDG logic manually for `Cache`, `Data` and `State`, at which point there is no reason to use `GetUserConfigDir` only for `Config`.

For now, `Cache` and `State` aren't used anywhere. If you have any suggestions regarding which other old `DataDir()` calls should be replaced with `Cache`, `Config` or `State`, feel free to point them out. I am not too familiar with the various audacity `xml` files, so I don't know, which of them should go to `Config` or `Data` or `State` etc.

**Comment by RuRo** [comment]
Here are some files/directories that currently still use `DataDir()`:
- lastlog.txt
- genres.txt
- journal.txt
- journallog.txt
- EQCurves.xml
- EQDefaultCurves.xml
- EQBackup.xml
- ffmpeg_presets.xml
- Chains
- Macros
- NRP
- Plug-Ins
- Theme

I might have missed some.

**Comment by RuRo** [comment]
Oops. I don't have a Windows/Mac machine, so I missed a typo. @crsib sorry, can you approve the CI/CD again.

**Comment by RuRo** [comment]
@imciner2 I've refactored the code to use `glib` and `get_user_*_dir()`, but it seems that `glib` is normally not used for building `libraries/` so I had to manually add `glib` to `libraries/lib-files/CMakeLists.txt`. I am not sure, if I did that right (I just copy-pasted that line from `src/CMakeLists.txt`).

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:29]: Looks like glib is included unconditionally.

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:252]: Let's use c-assert. We are moving away from wxASSERT

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:266]: Ditto

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:271]: You return non const reference from gTargetDir, but here C++ deduce the type as `FilePath`. Is is desired?

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:319]: Looks like you want overwrite the value, so you need `auto& dir = gTargetDir(target);`

**Review by crsib — CHANGES_REQUESTED** [review]
Glib is included and used unconditionally which is definitely a blocker. Some other issues as well.

**Comment by crsib** [comment]
Linux build has failed as well!

**Comment by RuRo** [comment]
> Linux build has failed as well!

Yeah. It seems that CI uses an older GLib version, that doesn't have `g_get_user_state_dir`. It seems to be only available since 2.71.1. What do you think we should do:

1) Update required GLib version
2) Remove `FileNames::StateDir` and `State` from `DirTarget`
3) Don't use GLib at all and revert to manually checking the environment variables

**Comment by crsib** [comment]
We will not upgrade GLib version on CI any time soon, as it will break AppImage for users on older GNU/Linux versions.

Otherwise it's your call

**Comment by RuRo** [comment]
@imciner2 requested to use GLib
@crsib requested to add `State`

Please decide, which request should be reverted. :sweat_smile:
I would generally prefer to just remove the GLib dependency and revert to manual env lookup, but this is a weak preference.

**Comment by crsib** [comment]
>I would generally prefer to just remove the GLib dependency and revert to manual env lookup,

Me too, to be fair.

**Comment by imciner2** [comment]
If you really want to use the state directory, then more work would be needed probably. The XDG state directory was only added to the spec last year, so there may be issues with the base directories not existing on older systems, so you need to ensure the entire directory path exists, not just the Audacity subdirectory.

> We will not upgrade GLib version on CI any time soon, as it will break AppImage for users on older GNU/Linux versions.

Doesn't appimage bundle its dependencies into the image?

**Comment by RuRo** [comment]
> ... there may be issues with the base directories not existing on older systems, so you need to ensure the entire directory path exists, not just the Audacity subdirectory. ...

Currently, the code calls `FileNames::MkDir` that uses `wxFileName::Mkdir` with `wxPATH_MKDIR_FULL`
```
If the flags contain wxPATH_MKDIR_FULL flag, try to create each directory in the path and also don't return an error if the target directory already exists.
```

So that shouldn't be an issue.

**Comment by RuRo** [comment]
So, @crsib @imciner2 what's the conclusion here? Is there anything else I can do to get this PR accepted?

**Comment by crsib** [comment]
I'll do a hopefully final round today

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:51]: I think we can do something like

```
namespace
{
auto GetOldUserDataDir()
{
    static const auto dir = wxFileName::GetHomeDir() + wxT("/.audacity-data");
    return dir;
}

auto OldUserDataDirExists()
{
    static const auto exists = wxDirExists(GetOldUserDataDir());
    return exists;
}
}
```

This will simplify code below

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:54]: I think we can have an array here by adding `Count` to  `DirTarget`:

```
enum class DirTarget
{
    Cache,
    Config,
    Data,
    State,

     Count
};

wxString directories[size_t(DirTarget::Count)];
```

Comment (crsib) [reply libraries/lib-files/FileNames.cpp:54]: (Sorry for the formatting, GitHub aint a good text editor xD)

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:244]: Can be done as
```
struct XDGDirConfig final
{
   wxString envName;
   wxString defaultDir;
};

XDGDirConfig xdgConfig[] = {
  { L"XDG_CACHE_HOME", L"/.cache" },
...
};

static_assert(sizeof(xdgConfig) / sizeof(xdgConfig[0] == size_t(DirTarget::Count)

wxString GetXDGDir(DirTarget target)
{
   const auto& dirConfig = xdgConfig[size_t(target)];
 
   wxString dir;
   // see if the XDG_*_HOME env var is defined. if it is, use its value.
   // if it isn't, use the default XDG-specified value.
   if ( !wxGetEnv(dirConfig.envName, &dir) || dir.empty() )
        dir = wxFileName::GetHomeDir() +  dirConfig.defaultDir;

  return dir;
}
```

This way we force the compiler to do all the checks necessary. Given that the enum is private, there is not much need left for any runtime checks.

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:291]: Let's name it `GetUserTargetDir` for consistency please!

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:293]: This becomes `auto& dir = directories[size_t(target)];`

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:314]: This goes away

File: libraries/lib-files/FileNames.cpp
Comment (crsib) [inline libraries/lib-files/FileNames.cpp:323]: `if (OldUserDataDirExists())...`

File: libraries/lib-files/FileNames.h
Comment (crsib) [inline libraries/lib-files/FileNames.h:117]: Let's make all 4 functions to be const!

**Review by crsib — CHANGES_REQUESTED** [review]
A small suggested rewrite :-)

File: libraries/lib-files/FileNames.cpp
Comment (RuRo) [reply libraries/lib-files/FileNames.cpp:51]: Actually, I think I'll just extract all the XDG-specific logic into a separate function and then just inline these 2 functions into it.

File: libraries/lib-files/FileNames.cpp
Comment (RuRo) [reply libraries/lib-files/FileNames.cpp:54]: Makes sense. Would `_targetCount` be an acceptable name? I would like to somehow distinguish that `Count` is a sentinel, not a valid Target.

File: libraries/lib-files/FileNames.cpp
Comment (RuRo) [reply libraries/lib-files/FileNames.cpp:244]: How about
```c++
using XDGDirConfig = std::tuple<...>;
```
Also I think the `xdgConfig` array should be `const`.

Comment (crsib) [reply libraries/lib-files/FileNames.cpp:244]: std::pair would do too, but I prefer descriptive names :-)

File: libraries/lib-files/FileNames.h
Comment (RuRo) [reply libraries/lib-files/FileNames.h:117]: What do you mean? This is a namespace, not a class.

Comment (crsib) [reply libraries/lib-files/FileNames.h:117]: Oh shoot, ignore me than

**Comment by RuRo** [comment]
@crsib can you approve the CI again, please.

**Review by crsib — APPROVED** [review]

**Pull request merged**
